File: navigator.py
import os
import sys
sys.path.append(os.path.dirname(os.path.realpath(sys.argv[0])) + "/lib/grph-1.5.24.jar")
from sikuli import *
from grph.in_memory import InMemoryGrph
from enum import Enum

class Navigator:
    def __init__(self, current_scene = "port"):
        self.graph = InMemoryGrph()
        self.actions = {}
        self.current_scene = current_scene
    def add_path(self, from_scene, to_scene, action):
        pair = from_scene.value, to_scene.value
        self.graph.addDirectedSimpleEdge(from_scene.value, to_scene.value)
        self.actions[pair] = action
    def navigate_to(self, target_scene):
        if current_scene == target_scene:
            return
        path = self.graph.getShortestPath( self.current_scene.value, target_scene.value).toVertexArray()
        for edge in zip(path[:-1], path):
            self.action[edge]()

Scene = Enum("Scene", "hokyu port fixing")

navigator = Navigator()

def scene(scene):
    def decorator(f):
        navigator.navigate_to(scene)
        f()
    return decorator

def click_port_button():
    while exists("1413605460003.png", 0):
        click(Location(10, 10))

for scene in [Scene.hokyu, Scene.fixing]:
    navigator.add_path(scene, Scene.port, click_port_button)

def click_and_wait_sidebar(button):
    def nav():
        click(button)
        wait("1413605460003.png")
    return nav

navigator.add_path(Scene.port, Scene.hokyu, click_and_wait_sidebar("1413617134671.png"))

File: navigator.rb
java_import 'grph.in_memory.InMemoryGrph'
require 'sikulix'
include Sikulix
require 'set'

class Navigator
  def initialize(current_scene = :port)
    @graph = InMemoryGrph.new
    @handlers = Set.new
    @actions = Hash.new { |hash, key| hash[key] =  Hash.new(Proc.new) }
    @current_scene = current_scene
  end

  def on_move(&handler)
    @handlers.add handler
  end

  def set_action(*edge, &action)
    source_id, target_id = ids = edge.map(&:object_id)
    @graph.addDirectedSimpleEdge(*ids) unless @graph.isDirectedSimpleEdge(*ids)
    @actions[source_id][target_id] = action
  end

  def navigate_to(target_scene)
    current_scene_id = @current_scene.object_id
    target_scene_id = target_scene.object_id
    if current_scene_id == target_scene_id
      @actions[current_scene_id][target_scene_id].call
    elsif path = @graph.getShortestPath(current_scene_id, target_scene_id)
      path[1..-1].zip(path) do |(target, source)|
        @actions[source][target].call
        @handlers.each do |handler|
          handler.call(ObjectSpace._id2ref source, ObjectSpace._id2ref target)
        end
      end
      @current_scene = target_scene
    else
      fail "Cannot find a path from #{@current_scene} to #{target_scene}"
    end
  end
end

navigator = Navigator.new

def scene(scene)
  navigator.navigate_to scene
  result = yield
  navigator.navigate_to scene
  result
end

[:hokyu, :fixing, :shutsugeki].each do |scene|
  navigator.set_action scene, :port do
    click Location.new(10, 10) while exists 'images/sidebar/port.png', 0
  end
end

[:hokyu, :fixing, :shutsugeki].each do |scene|
  navigator.set_action :port, scene do
    click "images/port/menu/#{scene}.png"
    wait 'images/sidebar/port.png'
  end
end